### DISCUSSION TITLE
使用R包texteffect发现文本中的潜在处理变量并估计其因果效应

### DISCUSSION THREAD

**Discussion by chengjun** [chengjun/mybook#46] [General]
[![苏格拉底大王](https://img1.doubanio.com/icon/up1558440-49.jpg)](https://www.douban.com/people/frankwang/)

texteffect: Discovering Latent Treatments in Text Corpora and Estimating Their Causal Effects
---------------------------------------------------------------------------------------------

> Fong and Grimmer (2016) Discovery of Treatments from Text Corpora. Proceedings of the 54th Annual Meeting of the Association for Computational Linguistics, pages 1600–1609, Berlin, Germany, August 7-12, 2016 <https://aclweb.org/anthology/P/P16/P16-1151.pdf>

Extensive literature in computational social science examines how features of messages, advertisements, and other corpora affect individuals’ decisions, but these analyses must specify the relevant features of the text before the experiment.

* Automated text analysis methods are able to discover features of text, but these methods cannot be used to obtain the estimates of causal effects—the quantity of interest for applied researchers.
* We introduce a new experimental design and statistical model to simultaneously discover treatments in corpora and estimate causal effects for these discovered treatments.
* We prove the conditions to identify the treatment effects of texts and introduce the supervised Indian Buffet process to discover those treatments.
* Our method enables us to discover treatments in a training set using a collection of texts and individuals’ responses to those texts, and then estimate the effects of these interventions in a test set of new texts and survey respondents.
* We apply the model to an experiment about candidate biographies, recovering intuitive features of voters’ decisions and revealing a penalty for lawyers and a bonus for military service.

![](https://img2.doubanio.com/view/group_topic/l/public/p583713901.webp)

Implements the approach described in Fong and Grimmer (2016) for automatically discovering latent treatments from a corpus and estimating the average marginal component effect (AMCE) of each treatment.

* The data is divided into a training and test set.

The supervised Indian Buffet Process (sibp) is used to discover latent treatments in the training set.

The fitted model is then applied to the test set to infer the values of the latent treatments in the test set.

Finally, Y is regressed on the latent treatments in the test set to estimate the causal effect of each treatment.

![](https://img2.doubanio.com/view/group_topic/l/public/p583714621.webp)

![](https://img2.doubanio.com/view/group_topic/l/public/p583714791.webp)

[CRAN - Package texteffect](https://cran.r-project.org/web/packages/texteffect/index.html)

```r
library(texteffect)
```
Load the Wikipedia biography data
---------------------------------

A data frame consisting of 51 columns (including an outcome measure and counts for each word in a 50 word vocabulary) and 250 observations.

```r
data(BioSample)

head(BioSample)
```

![](https://img9.doubanio.com/view/group_topic/l/public/p583717675.webp)

# Divide into training and test sets

```r
Y <- BioSample[,1]

X <- BioSample[,-1]

set.seed(1)

train.ind <- sample(1:nrow(X), size = 0.5*nrow(X), replace = FALSE)

# Fit an sIBP on the training data

sibp.fit <- sibp(X, Y, K = 2, alpha = 4, sigmasq.n = 0.8,

train.ind = train.ind)
```

sibp_amce Infer Treatments on the Test Set¶
-------------------------------------------

A table where the first column is the index of the treatment, the second column ("effect") is the estimated AMCE, the third column ("L") is the lower bound of the confidence interval, and the fourth column ("U") is the upper bound of the confidence interval.

```r
sibp.amce <- sibp_amce(sibp.fit, X, Y)

sibp.amce
```

![](https://img9.doubanio.com/view/group_topic/l/public/p583717855.webp)

sibp_amce_plot(sibp.amce)

![](https://img2.doubanio.com/view/group_topic/l/public/p583717901.webp)

# Infer the latent treatments in the test set

```r
infer_Z(sibp.fit, X)
```

![](https://img2.doubanio.com/view/group_topic/l/public/p583717931.webp)

```r
# Search sIBP for several parameter configurations; fit each to the training set
sibp.search <- sibp_param_search(X, Y, K = 2, alphas = c(2,4), sigmasq.ns = c(0.8, 1),

iters = 1, train.ind = train.ind)

# Get metric for evaluating most promising parameter configurations

sibp_rank_runs(sibp.search, X, 10)
```

![](https://img9.doubanio.com/view/group_topic/l/public/p583718016.webp)

# Qualitatively look at the top candidates
```r
sibp_top_words(sibp.search[["4"]][["0.8"]][[1]], colnames(X), 10, verbose = TRUE)
```

![](https://img2.doubanio.com/view/group_topic/l/public/p583718203.webp)

![](https://img9.doubanio.com/view/group_topic/l/public/p583718224.webp)
